File: Source/Nava.Runtime/ThrowHelpers.fs
namespace Nava.Runtime

open System
open System.Collections.Generic
open System.Diagnostics.CodeAnalysis
open System.Runtime.CompilerServices
open System.Runtime.Serialization

[<NoEquality; NoComparison; RequireQualifiedAccess>]
type ExceptionArgument =
    | Action
    | Array
    | ArrayIndex
    | CallBack
    | Capacity
    | Ch
    | Collection
    | Comparable
    | Comparer
    | Comparison
    | ComparisonType
    | Converter
    | Count
    | CreationOptions
    | Culture
    | Delay
    | DestinationArray
    | DestinationIndex
    | Dictionary
    | ElementType
    | EndIndex
    | Enumerable
    | Exception
    | Exceptions
    | Format
    | Function
    | Index
    | Indices
    | Info
    | Input
    | Item
    | Key
    | Keys
    | Len
    | Length
    | Lengths
    | List
    | LowerBounds
    | Manager
    | Match
    | MillisecondsDelay
    | MillisecondsTimeout
    | NewSize
    | Obj
    | Other
    | S
    | Source
    | SourceArray
    | SourceBytesToCopy
    | SourceIndex
    | Start
    | StartIndex
    | State
    | Task
    | Text
    | Timeout
    | Type
    | Value
    | Values

    override x.ToString() =
        match x with
        | Action -> "action"
        | Array -> "array"
        | ArrayIndex -> "arrayIndex"
        | CallBack -> "callBack"
        | Capacity -> "capacity"
        | Ch -> "ch"
        | Collection -> "collection"
        | Comparable -> "comparable"
        | Comparer -> "comparer"
        | Comparison -> "comparison"
        | ComparisonType -> "comparisonType"
        | Converter -> "converter"
        | Count -> "count"
        | CreationOptions -> "creationOptions"
        | Culture -> "culture"
        | Delay -> "delay"
        | DestinationArray -> "destinationArray"
        | DestinationIndex -> "destinationIndex"
        | Dictionary -> "dictionary"
        | ElementType -> "elementType"
        | EndIndex -> "endIndex"
        | Enumerable -> "enumerable"
        | Exception -> "exception"
        | Exceptions -> "exceptions"
        | Format -> "format"
        | Function -> "function"
        | Index -> "index"
        | Indices -> "indices"
        | Info -> "info"
        | Input -> "input"
        | Item -> "item"
        | Key -> "key"
        | Keys -> "keys"
        | Len -> "len"
        | Length -> "length"
        | Lengths -> "lengths"
        | List -> "list"
        | LowerBounds -> "lowerBounds"
        | Manager -> "manager"
        | Match -> "match"
        | MillisecondsDelay -> "millisecondsDelay"
        | MillisecondsTimeout -> "millisecondsTimeout"
        | NewSize -> "newSize"
        | Obj -> "obj"
        | Other -> "other"
        | S -> "s"
        | Source -> "source"
        | SourceArray -> "sourceArray"
        | SourceBytesToCopy -> "sourceBytesToCopy"
        | SourceIndex -> "sourceIndex"
        | Start -> "start"
        | StartIndex -> "startIndex"
        | State -> "state"
        | Task -> "task"
        | Text -> "text"
        | Timeout -> "timeout"
        | Type -> "type"
        | Value -> "value"
        | Values -> "values"

[<NoEquality; NoComparison>]
type ExceptionResource =
    | Argument_AddingDuplicate
    | Argument_ArrayPlusOffTooSmall
    | Argument_InvalidArgumentForComparison
    | Argument_InvalidOffLen
    | Argument_LowerBoundsMustMatch
    | Argument_MustBeType
    | Argument_NonZeroLowerBound
    | Argument_RankMultiDimNotSupported
    | Argument_TypeNotSupported
    | ArgumentException_OtherNotArrayOfCorrectLength
    | ArgumentOutOfRange_BiggerThanCollection
    | ArgumentOutOfRange_Count
    | ArgumentOutOfRange_EndIndexStartIndex
    | ArgumentOutOfRange_HugeArrayNotSupported
    | ArgumentOutOfRange_Index
    | ArgumentOutOfRange_ListInsert
    | ArgumentOutOfRange_NeedNonNegNum
    | ArgumentOutOfRange_SmallCapacity
    | ConcurrentCollection_SyncRoot_NotSupported
    | InvalidOperation_HSCapacityOverflow
    | InvalidOperation_IComparerFailed
    | InvalidOperation_NullArray
    | NotSupported_FixedSizeCollection
    | NotSupported_KeyCollectionSet
    | NotSupported_ReadOnlyCollection
    | NotSupported_StringComparison
    | NotSupported_ValueCollectionSet
    | Rank_MultiDimNotSupported
    | Serialization_MissingKeys
    | Serialization_NullKey

    override x.ToString() =
        match x with
        | Argument_AddingDuplicate -> "Duplicate item added."
        | Argument_ArrayPlusOffTooSmall -> "Array plus offset too small."
        | Argument_InvalidArgumentForComparison -> "Invalid argument for comparison."
        | Argument_InvalidOffLen -> "Invalid offset length."
        | Argument_LowerBoundsMustMatch -> "Array lower bounds must match."
        | Argument_MustBeType -> "Argument must be of type: "
        | Argument_NonZeroLowerBound -> "Arrays with a non-zero lower bound are not supported."
        | Argument_RankMultiDimNotSupported -> "Multi-dimensional arrays are not supported."
        | Argument_TypeNotSupported -> "Type not supported."
        | ArgumentException_OtherNotArrayOfCorrectLength -> "The other array is not of the correct length."
        | ArgumentOutOfRange_BiggerThanCollection -> "The given value was larger than the size of the collection."
        | ArgumentOutOfRange_Count -> "Argument 'count' was out of the range of valid values."
        | ArgumentOutOfRange_EndIndexStartIndex -> "The end index does not come after the start index."
        | ArgumentOutOfRange_HugeArrayNotSupported -> "Huge arrays are not supported."
        | ArgumentOutOfRange_Index -> "Argument 'index' was out of the range of valid values."
        | ArgumentOutOfRange_ListInsert -> "Insertion index was out of the range of valid values."
        | ArgumentOutOfRange_NeedNonNegNum -> "The number must be non-negative."
        | ArgumentOutOfRange_SmallCapacity -> "The capacity cannot be set below the current Count."
        | ConcurrentCollection_SyncRoot_NotSupported -> "SyncRoot not supported."
        | InvalidOperation_HSCapacityOverflow -> "Set hash capacity overflow. Cannot increase size."
        | InvalidOperation_IComparerFailed -> "IComparer failed."
        | InvalidOperation_NullArray -> "Null arrays are not supported."
        | NotSupported_FixedSizeCollection -> "This operation is not suppored on a fixed-size collection."
        | NotSupported_KeyCollectionSet -> "The KeyCollection does not support modification."
        | NotSupported_ReadOnlyCollection -> "This operation is not supported on a read-only collection."
        | NotSupported_StringComparison -> "String comparison not supported."
        | NotSupported_ValueCollectionSet -> "The ValueCollection does not support modification."
        | Rank_MultiDimNotSupported -> "Multi-dimensional arrays are not supported."
        | Serialization_MissingKeys -> "Serialization error: missing keys."
        | Serialization_NullKey -> "Serialization error: null key."

type ThrowHelpers =
    [<DoesNotReturn>]
    static member ThrowArrayTypeMismatchException() : unit = ArrayTypeMismatchException() |> raise

    [<DoesNotReturn>]
    static member ThrowIndexOutOfRangeException() : unit = IndexOutOfRangeException() |> raise

    [<DoesNotReturn>]
    static member ThrowArgumentOutOfRangeException() : unit = ArgumentOutOfRangeException() |> raise

    [<DoesNotReturn>]
    static member ThrowArgumentException_DestinationTooShort() : unit =
        ArgumentException("Destination too short.") |> raise

    [<DoesNotReturn>]
    static member ThrowArgumentException_OverlapAlignmentMismatch() : unit =
        ArgumentException("Overlap alignment mismatch.") |> raise

    [<DoesNotReturn>]
    static member ThrowArgumentOutOfRange_IndexException() : unit =
        ThrowHelpers.GetArgumentOutOfRangeException(ExceptionArgument.Index, ArgumentOutOfRange_Index)
        |> raise

    [<DoesNotReturn>]
    static member ThrowIndexArgumentOutOfRange_NeedNonNegNumException() : unit =
        ThrowHelpers.GetArgumentOutOfRangeException(ExceptionArgument.Index, ArgumentOutOfRange_NeedNonNegNum)
        |> raise

    [<DoesNotReturn>]
    static member ThrowValueArgumentOutOfRange_NeedNonNegNumException() : unit =
        ThrowHelpers.GetArgumentOutOfRangeException(ExceptionArgument.Value, ArgumentOutOfRange_NeedNonNegNum)
        |> raise

    [<DoesNotReturn>]
    static member ThrowLengthArgumentOutOfRange_ArgumentOutOfRange_NeedNonNegNum() : unit =
        ThrowHelpers.GetArgumentOutOfRangeException(ExceptionArgument.Length, ArgumentOutOfRange_NeedNonNegNum)
        |> raise

    [<DoesNotReturn>]
    static member ThrowStartIndexArgumentOutOfRange_ArgumentOutOfRange_Index() : unit =
        ThrowHelpers.GetArgumentOutOfRangeException(ExceptionArgument.StartIndex, ArgumentOutOfRange_Index)
        |> raise

    [<DoesNotReturn>]
    static member ThrowCountArgumentOutOfRange_ArgumentOutOfRange_Count() : unit =
        ThrowHelpers.GetArgumentOutOfRangeException(ExceptionArgument.Count, ArgumentOutOfRange_Count)
        |> raise

    [<DoesNotReturn>]
    static member ThrowWrongKeyTypeArgumentException<'T>(key: 'T, targetType: Type) : unit =
        // Generic key to move the boxing to the right hand side of throw
        ThrowHelpers.GetWrongKeyTypeArgumentException(key, targetType) |> raise

    [<DoesNotReturn>]
    static member ThrowWrongValueTypeArgumentException<'T>(value: 'T, targetType: Type) : unit =
        // Generic key to move the boxing to the right hand side of throw
        ThrowHelpers.GetWrongValueTypeArgumentException(value, targetType) |> raise

    [<DoesNotReturn>]
    static member GetAddingDuplicateWithKeyArgumentException(key: 'T) =
        let defaultInterpolatedStringHandler = DefaultInterpolatedStringHandler(34, 1)
        defaultInterpolatedStringHandler.AppendLiteral("Error adding duplicate with key: ")
        defaultInterpolatedStringHandler.AppendFormatted(key)
        defaultInterpolatedStringHandler.AppendLiteral(".")
        ArgumentException(defaultInterpolatedStringHandler.ToStringAndClear())

    [<DoesNotReturn>]
    static member ThrowAddingDuplicateWithKeyArgumentException<'T>(key: 'T) : unit =
        // Generic key to move the boxing to the right hand side of throw
        ThrowHelpers.GetAddingDuplicateWithKeyArgumentException(key) |> raise

    [<DoesNotReturn>]
    static member ThrowKeyNotFoundException(key: 'T) : unit =
        // Generic key to move the boxing to the right hand side of throw
        ThrowHelpers.GetKeyNotFoundException(key) |> raise

    [<DoesNotReturn>]
    static member ThrowArgumentException(resource: ExceptionResource) : unit =
        ThrowHelpers.GetArgumentException(resource) |> raise

    [<DoesNotReturn>]
    static member ThrowArgumentException(resource: ExceptionResource, argument: ExceptionArgument) : unit =
        ThrowHelpers.GetArgumentExceptionWithArgument(resource, argument) |> raise

    [<DoesNotReturn>]
    static member GetArgumentNullException(argument: ExceptionArgument) =
        ArgumentNullException(argument.ToString())

    [<DoesNotReturn>]
    static member ThrowArgumentNullException(argument: ExceptionArgument) : unit =
        ThrowHelpers.GetArgumentNullException(argument) |> raise

    [<DoesNotReturn>]
    static member ThrowArgumentNullException(resource: ExceptionResource) : unit =
        ArgumentNullException(resource.ToString()) |> raise

    [<DoesNotReturn>]
    static member ThrowArgumentNullException(argument: ExceptionArgument, resource: ExceptionResource) : unit =
        ArgumentNullException(argument.ToString(), resource.ToString()) |> raise

    [<DoesNotReturn>]
    static member ThrowArgumentOutOfRangeException(argument: ExceptionArgument) : unit =
        ArgumentOutOfRangeException(argument.ToString()) |> raise

    [<DoesNotReturn>]
    static member ThrowArgumentOutOfRangeException(argument: ExceptionArgument, resource: ExceptionResource) : unit =
        ThrowHelpers.GetArgumentOutOfRangeException(argument, resource) |> raise

    [<DoesNotReturn>]
    static member ThrowArgumentOutOfRangeException(argument: ExceptionArgument, paramNumber: int, resource: ExceptionResource) : unit =
        ThrowHelpers.GetArgumentOutOfRangeException(argument, paramNumber, resource)
        |> raise

    [<DoesNotReturn>]
    static member ThrowInvalidOperationException(resource: ExceptionResource) : unit =
        ThrowHelpers.GetInvalidOperationException(resource) |> raise

    [<DoesNotReturn>]
    static member ThrowInvalidOperationException(resource: ExceptionResource, e) : unit =
        InvalidOperationException(resource.ToString(), e) |> raise

    [<DoesNotReturn>]
    static member ThrowSerializationException(resource: ExceptionResource) : unit =
        SerializationException(resource.ToString()) |> raise

    [<DoesNotReturn>]
    static member ThrowSecurityException(resource: ExceptionResource) : unit =
        Security.SecurityException(resource.ToString()) |> raise

    [<DoesNotReturn>]
    static member ThrowRankException(resource: ExceptionResource) : unit =
        RankException(resource.ToString()) |> raise

    [<DoesNotReturn>]
    static member ThrowNotSupportedException(resource: ExceptionResource) : unit =
        NotSupportedException(resource.ToString()) |> raise

    [<DoesNotReturn>]
    static member ThrowUnauthorizedAccessException(resource: ExceptionResource) : unit =
        UnauthorizedAccessException(resource.ToString()) |> raise

    [<DoesNotReturn>]
    static member ThrowObjectDisposedException(objectName, resource: ExceptionResource) : unit =
        ObjectDisposedException(objectName, resource.ToString()) |> raise

    [<DoesNotReturn>]
    static member ThrowObjectDisposedException(resource: ExceptionResource) : unit =
        ObjectDisposedException(null, resource.ToString()) |> raise

    [<DoesNotReturn>]
    static member ThrowNotSupportedException() : unit = NotSupportedException() |> raise

    [<DoesNotReturn>]
    static member ThrowAggregateException(exceptions: List<Exception>) : unit = AggregateException(exceptions) |> raise

    [<DoesNotReturn>]
    static member ThrowOutOfMemoryException() : unit = OutOfMemoryException() |> raise

    [<DoesNotReturn>]
    static member ThrowArgumentException_Argument_InvalidArrayType() : unit =
        ArgumentException("Invalid array type.") |> raise

    [<DoesNotReturn>]
    static member ThrowInvalidOperationException_InvalidOperation_EnumNotStarted() : unit =
        InvalidOperationException("Enumeration has not started.") |> raise

    [<DoesNotReturn>]
    static member ThrowInvalidOperationException_InvalidOperation_EnumEnded() : unit =
        InvalidOperationException("Enumeration has ended.") |> raise

    [<DoesNotReturn>]
    static member ThrowInvalidOperationException_EnumCurrent(index: int) : unit =
        ThrowHelpers.GetInvalidOperationException_EnumCurrent(index) |> raise

    [<DoesNotReturn>]
    static member ThrowInvalidOperationException_InvalidOperation_EnumFailedVersion() : unit =
        InvalidOperationException("Collection was modified during enumeration.")
        |> raise

    [<DoesNotReturn>]
    static member ThrowInvalidOperationException_InvalidOperation_EnumOpCantHappen() : unit =
        InvalidOperationException("Invalid enumerator state: enumeration cannot proceed.")
        |> raise

    [<DoesNotReturn>]
    static member ThrowInvalidOperationException_InvalidOperation_NoValue() : unit =
        InvalidOperationException("No value provided.") |> raise

    [<DoesNotReturn>]
    static member ThrowInvalidOperationException_ConcurrentOperationsNotSupported() : unit =
        InvalidOperationException("Concurrent operations are not supported.") |> raise

    [<DoesNotReturn>]
    static member ThrowInvalidOperationException_HandleIsNotInitialized() : unit =
        InvalidOperationException("Handle is not initialized.") |> raise

    [<DoesNotReturn>]
    static member ThrowFormatException_BadFormatSpecifier() : unit =
        FormatException("Bad format specifier.") |> raise

    [<DoesNotReturn>]
    static member GetArgumentException(resource: ExceptionResource) = ArgumentException(resource.ToString())

    [<DoesNotReturn>]
    static member GetInvalidOperationException(resource: ExceptionResource) =
        InvalidOperationException(resource.ToString())

    static member GetWrongKeyTypeArgumentException(key: 'T, targetType: Type) =
        let defaultInterpolatedStringHandler = DefaultInterpolatedStringHandler(35, 2)
        defaultInterpolatedStringHandler.AppendLiteral("Wrong key type. Expected ")
        defaultInterpolatedStringHandler.AppendFormatted(targetType)
        defaultInterpolatedStringHandler.AppendLiteral(", got: '")
        defaultInterpolatedStringHandler.AppendFormatted(key)
        defaultInterpolatedStringHandler.AppendLiteral("'.")
        ArgumentException(defaultInterpolatedStringHandler.ToStringAndClear(), "key")

    static member GetWrongValueTypeArgumentException(value: 'T, targetType: Type) =
        let defaultInterpolatedStringHandler = DefaultInterpolatedStringHandler(37, 2)
        defaultInterpolatedStringHandler.AppendLiteral("Wrong value type. Expected ")
        defaultInterpolatedStringHandler.AppendFormatted(targetType)
        defaultInterpolatedStringHandler.AppendLiteral(", got: '")
        defaultInterpolatedStringHandler.AppendFormatted(value)
        defaultInterpolatedStringHandler.AppendLiteral("'.")
        ArgumentException(defaultInterpolatedStringHandler.ToStringAndClear(), "value")

    static member GetKeyNotFoundException(key: 'T) =
        let defaultInterpolatedStringHandler = DefaultInterpolatedStringHandler(15, 1)
        defaultInterpolatedStringHandler.AppendLiteral("Key not found: ")
        defaultInterpolatedStringHandler.AppendFormatted(key)
        KeyNotFoundException(defaultInterpolatedStringHandler.ToStringAndClear())

    static member GetArgumentOutOfRangeException(argument: ExceptionArgument, resource: ExceptionResource) =
        ArgumentOutOfRangeException(argument.ToString(), resource.ToString())

    static member GetArgumentExceptionWithArgument(resource: ExceptionResource, argument: ExceptionArgument) =
        ArgumentException(resource.ToString(), argument.ToString())

    static member GetArgumentOutOfRangeException(argument: ExceptionArgument, paramNumber: int, resource: ExceptionResource) =
        ArgumentOutOfRangeException(argument.ToString() + "[" + paramNumber.ToString() + "]", resource.ToString())

    static member GetInvalidOperationException_EnumCurrent(index: int) =
        InvalidOperationException(
            if index < 0 then
                "Enumeration has not started"
            else
                "Enumeration has ended"
        )

    // Allow nulls for reference types and Nullable<U>, but not for value types.
    // Aggressively inline so the jit evaluates the if in place and either drops the call altogether
    // Or just leaves null test and call to the Non-returning ThrowHelper.ThrowArgumentNullException
    [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
    static member inline IfNullAndNullsAreIllegalThenThrow<'T>(value: obj, argName: ExceptionArgument) : unit =
        // Note that default<'T> is not equal to null for value types except when T is Nullable<U>.
        if Object.ReferenceEquals(Unchecked.defaultof<'T>, null) |> not && value |> isNull then
            ThrowHelpers.ThrowArgumentNullException(argName)

    [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
    static member inline ThrowForUnsupportedVectorBaseType<'T when 'T: struct>() : unit =
        //To work around F# Equality. Generated code is closer to C#.
        if
            System.Type.op_Inequality(typeof<'T>, typeof<byte>)
            && System.Type.op_Inequality(typeof<'T>, typeof<sbyte>)
            && System.Type.op_Inequality(typeof<'T>, typeof<int16>)
            && System.Type.op_Inequality(typeof<'T>, typeof<uint16>)
            && System.Type.op_Inequality(typeof<'T>, typeof<int>)
            && System.Type.op_Inequality(typeof<'T>, typeof<uint>)
            && System.Type.op_Inequality(typeof<'T>, typeof<int64>)
            && System.Type.op_Inequality(typeof<'T>, typeof<uint64>)
            && System.Type.op_Inequality(typeof<'T>, typeof<float>)
            && System.Type.op_Inequality(typeof<'T>, typeof<double>)
        then
            ThrowHelpers.ThrowNotSupportedException(Argument_TypeNotSupported)

File: Source/Nava.Runtime.CS/ThrowHelpers.cs
using System.Diagnostics;

namespace Nava.Runtime;

//
// The convention for this enum is using the argument name as the enum name
//
internal enum ExceptionArgument {
	action,
	array,
	arrayIndex,
	callBack,
	capacity,
	ch,
	collection,
	comparable,
	comparer,
	comparison,
	comparisonType,
	converter,
	count,
	creationOptions,
	culture,
	delay,
	destinationArray,
	destinationIndex,
	dictionary,
	elementType,
	endIndex,
	enumerable,
	exception,
	exceptions,
	format,
	function,
	index,
	indices,
	info,
	input,
	item,
	key,
	keys,
	len,
	length,
	lengths,
	list,
	lowerBounds,
	manager,
	match,
	millisecondsDelay,
	millisecondsTimeout,
	newSize,
	obj,
	other,
	s,
	source,
	sourceArray,
	sourceBytesToCopy,
	sourceIndex,
	start,
	startIndex,
	state,
	task,
	text,
	timeout,
	type,
	value,
	values,
}

//
// The convention for this enum is using the resource name as the enum name
//
internal enum ExceptionResource {
	Arg_ArrayPlusOffTooSmall,
	Arg_LowerBoundsMustMatch,
	Arg_MustBeType,
	Arg_NonZeroLowerBound,
	Arg_RankMultiDimNotSupported,
	Arg_TypeNotSupported,
	Argument_AddingDuplicate,
	Argument_InvalidArgumentForComparison,
	Argument_InvalidOffLen,
	ArgumentException_OtherNotArrayOfCorrectLength,
	ArgumentOutOfRange_BiggerThanCollection,
	ArgumentOutOfRange_Count,
	ArgumentOutOfRange_EndIndexStartIndex,
	ArgumentOutOfRange_HugeArrayNotSupported,
	ArgumentOutOfRange_Index,
	ArgumentOutOfRange_ListInsert,
	ArgumentOutOfRange_NeedNonNegNum,
	ArgumentOutOfRange_SmallCapacity,
	ConcurrentCollection_SyncRoot_NotSupported,
	InvalidOperation_HSCapacityOverflow,
	InvalidOperation_IComparerFailed,
	InvalidOperation_NullArray,
	NotSupported_FixedSizeCollection,
	NotSupported_KeyCollectionSet,
	NotSupported_ReadOnlyCollection,
	NotSupported_StringComparison,
	NotSupported_ValueCollectionSet,
	Rank_MultiDimNotSupported,
	Serialization_MissingKeys,
	Serialization_NullKey,
}
internal static class ThrowHelpers {

	// Allow nulls for reference types and Nullable<U>, but not for value types.
	// Aggressively inline so the jit evaluates the if in place and either drops the call altogether
	// Or just leaves null test and call to the Non-returning ThrowHelper.ThrowArgumentNullException
	[MethodImpl(MethodImplOptions.AggressiveInlining)]
	internal static void IfNullAndNullsAreIllegalThenThrow<T>(object value, ExceptionArgument argName) {
		// Note that default(T) is not equal to null for value types except when T is Nullable<U>.
		if(!(default(T) == null) && value == null)
			ThrowHelpers.ThrowArgumentNullException(argName);
	}

	internal static void ThrowAddingDuplicateWithKeyArgumentException<T>(T key) => throw GetAddingDuplicateWithKeyArgumentException(key!);

	internal static void ThrowAggregateException(List<Exception> exceptions) => throw new AggregateException(exceptions);

	internal static void ThrowArgumentException(ExceptionResource resource) => throw GetArgumentException(resource);

	internal static void ThrowArgumentException(ExceptionResource resource, ExceptionArgument argument) => throw GetArgumentException(resource, argument);

	internal static void ThrowArgumentException_Argument_InvalidArrayType() => throw new ArgumentException("Invalid array type.");

	internal static void ThrowArgumentException_DestinationTooShort() => throw new ArgumentException("Destination too short.");

	internal static void ThrowArgumentException_OverlapAlignmentMismatch() => throw new ArgumentException("Overlap alignment mismatch.");

	internal static void ThrowArgumentNullException(ExceptionArgument argument) => throw GetArgumentNullException(argument);

	internal static void ThrowArgumentNullException(ExceptionResource resource) => throw new ArgumentNullException(GetResourceString(resource));

	internal static void ThrowArgumentNullException(ExceptionArgument argument, ExceptionResource resource) => throw new ArgumentNullException(GetArgumentName(argument), GetResourceString(resource));

	internal static void ThrowArgumentOutOfRangeException() => throw new ArgumentOutOfRangeException();

	internal static void ThrowArgumentOutOfRangeException(ExceptionArgument argument) => throw new ArgumentOutOfRangeException(GetArgumentName(argument));

	internal static void ThrowArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource) => throw GetArgumentOutOfRangeException(argument, resource);

	internal static void ThrowArgumentOutOfRangeException(ExceptionArgument argument, int paramNumber, ExceptionResource resource) => throw GetArgumentOutOfRangeException(argument, paramNumber, resource);

	internal static void ThrowArgumentOutOfRange_IndexException() => throw GetArgumentOutOfRangeException(ExceptionArgument.index,
												ExceptionResource.ArgumentOutOfRange_Index);
	internal static void ThrowArrayTypeMismatchException() => throw new ArrayTypeMismatchException();

	internal static void ThrowCountArgumentOutOfRange_ArgumentOutOfRange_Count() => throw GetArgumentOutOfRangeException(ExceptionArgument.count,
												ExceptionResource.ArgumentOutOfRange_Count);

	[MethodImpl(MethodImplOptions.AggressiveInlining)]
	internal static void ThrowForUnsupportedVectorBaseType<T>() where T : struct {
		if(typeof(T) != typeof(byte) && typeof(T) != typeof(sbyte) &&
			typeof(T) != typeof(short) && typeof(T) != typeof(ushort) &&
			typeof(T) != typeof(int) && typeof(T) != typeof(uint) &&
			typeof(T) != typeof(long) && typeof(T) != typeof(ulong) &&
			typeof(T) != typeof(float) && typeof(T) != typeof(double)) {
			ThrowNotSupportedException(ExceptionResource.Arg_TypeNotSupported);
		}
	}

	internal static void ThrowFormatException_BadFormatSpecifier() => throw new FormatException("Bad format specifier.");

	internal static void ThrowIndexArgumentOutOfRange_NeedNonNegNumException() => throw GetArgumentOutOfRangeException(ExceptionArgument.index,
												ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);

	internal static void ThrowIndexOutOfRangeException() => throw new IndexOutOfRangeException();

	internal static void ThrowInvalidOperationException(ExceptionResource resource) => throw GetInvalidOperationException(resource);

	internal static void ThrowInvalidOperationException(ExceptionResource resource, Exception e) => throw new InvalidOperationException(GetResourceString(resource), e);

	internal static void ThrowInvalidOperationException_ConcurrentOperationsNotSupported() => throw new InvalidOperationException("Concurrent operations are not supported.");

	internal static void ThrowInvalidOperationException_EnumCurrent(int index) => throw GetInvalidOperationException_EnumCurrent(index);

	internal static void ThrowInvalidOperationException_HandleIsNotInitialized() => throw new InvalidOperationException("Handle is not initialized.");

	internal static void ThrowInvalidOperationException_InvalidOperation_EnumEnded() => throw new InvalidOperationException("Enumeration has ended.");

	internal static void ThrowInvalidOperationException_InvalidOperation_EnumFailedVersion() => throw new InvalidOperationException("Collection was modified during enumeration.");

	internal static void ThrowInvalidOperationException_InvalidOperation_EnumNotStarted() => throw new InvalidOperationException("Enumeration has not started.");

	internal static void ThrowInvalidOperationException_InvalidOperation_EnumOpCantHappen() => throw new InvalidOperationException("Invalid enumerator state: enumeration cannot proceed.");

	internal static void ThrowInvalidOperationException_InvalidOperation_NoValue() => throw new InvalidOperationException("No value provided.");

	internal static void ThrowKeyNotFoundException<T>(T key) => throw GetKeyNotFoundException(key!);

	internal static void ThrowLengthArgumentOutOfRange_ArgumentOutOfRange_NeedNonNegNum() => throw GetArgumentOutOfRangeException(ExceptionArgument.length,
												ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);

	internal static void ThrowNotSupportedException(ExceptionResource resource) => throw new NotSupportedException(GetResourceString(resource));

	internal static void ThrowNotSupportedException() => throw new NotSupportedException();

	internal static void ThrowObjectDisposedException(string objectName, ExceptionResource resource) => throw new ObjectDisposedException(objectName, GetResourceString(resource));

	internal static void ThrowObjectDisposedException(ExceptionResource resource) => throw new ObjectDisposedException(null, GetResourceString(resource));

	internal static void ThrowOutOfMemoryException() => throw new OutOfMemoryException();

	internal static void ThrowRankException(ExceptionResource resource) => throw new RankException(GetResourceString(resource));

	internal static void ThrowSecurityException(ExceptionResource resource) => throw new System.Security.SecurityException(GetResourceString(resource));

	internal static void ThrowSerializationException(ExceptionResource resource) => throw new System.Runtime.Serialization.SerializationException(GetResourceString(resource));

	internal static void ThrowStartIndexArgumentOutOfRange_ArgumentOutOfRange_Index() => throw GetArgumentOutOfRangeException(ExceptionArgument.startIndex,
												ExceptionResource.ArgumentOutOfRange_Index);

	internal static void ThrowUnauthorizedAccessException(ExceptionResource resource) => throw new UnauthorizedAccessException(GetResourceString(resource));

	internal static void ThrowValueArgumentOutOfRange_NeedNonNegNumException() => throw GetArgumentOutOfRangeException(ExceptionArgument.value,
												ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);

	internal static void ThrowWrongKeyTypeArgumentException<T>(T key, Type targetType) => throw GetWrongKeyTypeArgumentException(key!, targetType);

	internal static void ThrowWrongValueTypeArgumentException<T>(T value, Type targetType) => throw GetWrongValueTypeArgumentException(value!, targetType);

	private static ArgumentException GetAddingDuplicateWithKeyArgumentException(object key) => new($"Error adding duplicate with key: {key}.");

	private static ArgumentException GetArgumentException(ExceptionResource resource) => new(GetResourceString(resource));

	private static ArgumentException GetArgumentException(ExceptionResource resource, ExceptionArgument argument) => new(GetResourceString(resource), GetArgumentName(argument));

	private static string GetArgumentName(ExceptionArgument argument) {
		return argument switch {
			ExceptionArgument.obj => "obj",
			ExceptionArgument.dictionary => "dictionary",
			ExceptionArgument.array => "array",
			ExceptionArgument.info => "info",
			ExceptionArgument.key => "key",
			ExceptionArgument.text => "text",
			ExceptionArgument.values => "values",
			ExceptionArgument.value => "value",
			ExceptionArgument.startIndex => "startIndex",
			ExceptionArgument.task => "task",
			ExceptionArgument.ch => "ch",
			ExceptionArgument.s => "s",
			ExceptionArgument.input => "input",
			ExceptionArgument.list => "list",
			ExceptionArgument.index => "index",
			ExceptionArgument.capacity => "capacity",
			ExceptionArgument.collection => "collection",
			ExceptionArgument.item => "item",
			ExceptionArgument.converter => "converter",
			ExceptionArgument.match => "match",
			ExceptionArgument.count => "count",
			ExceptionArgument.action => "action",
			ExceptionArgument.comparison => "comparison",
			ExceptionArgument.exceptions => "exceptions",
			ExceptionArgument.exception => "exception",
			ExceptionArgument.enumerable => "enumerable",
			ExceptionArgument.start => "start",
			ExceptionArgument.format => "format",
			ExceptionArgument.culture => "culture",
			ExceptionArgument.comparer => "comparer",
			ExceptionArgument.comparable => "comparable",
			ExceptionArgument.source => "source",
			ExceptionArgument.state => "state",
			ExceptionArgument.length => "length",
			ExceptionArgument.comparisonType => "comparisonType",
			ExceptionArgument.manager => "manager",
			ExceptionArgument.sourceBytesToCopy => "sourceBytesToCopy",
			ExceptionArgument.callBack => "callBack",
			ExceptionArgument.creationOptions => "creationOptions",
			ExceptionArgument.function => "function",
			ExceptionArgument.delay => "delay",
			ExceptionArgument.millisecondsDelay => "millisecondsDelay",
			ExceptionArgument.millisecondsTimeout => "millisecondsTimeout",
			ExceptionArgument.timeout => "timeout",
			ExceptionArgument.type => "type",
			ExceptionArgument.sourceIndex => "sourceIndex",
			ExceptionArgument.sourceArray => "sourceArray",
			ExceptionArgument.destinationIndex => "destinationIndex",
			ExceptionArgument.destinationArray => "destinationArray",
			ExceptionArgument.other => "other",
			ExceptionArgument.newSize => "newSize",
			ExceptionArgument.lowerBounds => "lowerBounds",
			ExceptionArgument.lengths => "lengths",
			ExceptionArgument.len => "len",
			ExceptionArgument.keys => "keys",
			ExceptionArgument.indices => "indices",
			ExceptionArgument.endIndex => "endIndex",
			ExceptionArgument.elementType => "elementType",
			ExceptionArgument.arrayIndex => "arrayIndex",
			_ => argument.ToString(),
		};
	}

	private static ArgumentNullException GetArgumentNullException(ExceptionArgument argument) => new(GetArgumentName(argument));

	private static ArgumentOutOfRangeException GetArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource) => new(GetArgumentName(argument), GetResourceString(resource));

	private static ArgumentOutOfRangeException GetArgumentOutOfRangeException(ExceptionArgument argument, int paramNumber, ExceptionResource resource) => new(GetArgumentName(argument) + "[" + paramNumber.ToString() + "]", GetResourceString(resource));

	private static InvalidOperationException GetInvalidOperationException(ExceptionResource resource) => new(GetResourceString(resource));

	private static InvalidOperationException GetInvalidOperationException_EnumCurrent(int index) => new(index < 0 ? "Enumeration has not started" : "Enumeration has ended");

	private static KeyNotFoundException GetKeyNotFoundException(object key) => new($"Key not found: {key}");

	private static string GetResourceString(ExceptionResource resource) {
		return resource switch {
			ExceptionResource.ArgumentOutOfRange_Index => "Argument 'index' was out of the range of valid values.",
			ExceptionResource.ArgumentOutOfRange_Count => "Argument 'count' was out of the range of valid values.",
			ExceptionResource.Arg_ArrayPlusOffTooSmall => "Array plus offset too small.",
			ExceptionResource.NotSupported_ReadOnlyCollection => "This operation is not supported on a read-only collection.",
			ExceptionResource.Arg_RankMultiDimNotSupported => "Multi-dimensional arrays are not supported.",
			ExceptionResource.Arg_NonZeroLowerBound => "Arrays with a non-zero lower bound are not supported.",
			ExceptionResource.ArgumentOutOfRange_ListInsert => "Insertion index was out of the range of valid values.",
			ExceptionResource.ArgumentOutOfRange_NeedNonNegNum => "The number must be non-negative.",
			ExceptionResource.ArgumentOutOfRange_SmallCapacity => "The capacity cannot be set below the current Count.",
			ExceptionResource.Argument_InvalidOffLen => "Invalid offset length.",
			ExceptionResource.ArgumentOutOfRange_BiggerThanCollection => "The given value was larger than the size of the collection.",
			ExceptionResource.Serialization_MissingKeys => "Serialization error: missing keys.",
			ExceptionResource.Serialization_NullKey => "Serialization error: null key.",
			ExceptionResource.NotSupported_KeyCollectionSet => "The KeyCollection does not support modification.",
			ExceptionResource.NotSupported_ValueCollectionSet => "The ValueCollection does not support modification.",
			ExceptionResource.InvalidOperation_NullArray => "Null arrays are not supported.",
			ExceptionResource.InvalidOperation_HSCapacityOverflow => "Set hash capacity overflow. Cannot increase size.",
			ExceptionResource.NotSupported_StringComparison => "String comparison not supported.",
			ExceptionResource.ConcurrentCollection_SyncRoot_NotSupported => "SyncRoot not supported.",
			ExceptionResource.ArgumentException_OtherNotArrayOfCorrectLength => "The other array is not of the correct length.",
			ExceptionResource.ArgumentOutOfRange_EndIndexStartIndex => "The end index does not come after the start index.",
			ExceptionResource.ArgumentOutOfRange_HugeArrayNotSupported => "Huge arrays are not supported.",
			ExceptionResource.Argument_AddingDuplicate => "Duplicate item added.",
			ExceptionResource.Argument_InvalidArgumentForComparison => "Invalid argument for comparison.",
			ExceptionResource.Arg_LowerBoundsMustMatch => "Array lower bounds must match.",
			ExceptionResource.Arg_MustBeType => "Argument must be of type: ",
			ExceptionResource.InvalidOperation_IComparerFailed => "IComparer failed.",
			ExceptionResource.NotSupported_FixedSizeCollection => "This operation is not suppored on a fixed-size collection.",
			ExceptionResource.Rank_MultiDimNotSupported => "Multi-dimensional arrays are not supported.",
			ExceptionResource.Arg_TypeNotSupported => "Type not supported.",
			_ => resource.ToString(),
		};
	}

	private static ArgumentException GetWrongKeyTypeArgumentException(object key, Type targetType) => new($"Wrong key type. Expected {targetType}, got: '{key}'.", nameof(key));

	private static ArgumentException GetWrongValueTypeArgumentException(object value, Type targetType) => new($"Wrong value type. Expected {targetType}, got: '{value}'.", nameof(value));
}
